core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -259,46 +257,42 @@ public void retry(String workflowId) throws Exception {
-        // update for these old tasks.
-        List<Task> update = workflow.getTasks().stream()
-                .filter(task -> !task.isRetried())
-                .collect(Collectors.toList());
-
-        update.forEach(task -> task.setRetried(true));
-        executionDAO.updateTasks(update);
-
-        List<Task> rescheduledTasks = new ArrayList<Task>();
+        List<Task> rescheduledTasks = new ArrayList<>();
         // Now reschedule the failed task
-        Task retried = failedTask.copy();
-        retried.setTaskId(IDGenerator.generate());
-        retried.setRetriedTaskId(failedTask.getTaskId());
-        retried.setStatus(SCHEDULED);
-        retried.setRetryCount(failedTask.getRetryCount() + 1);
-        rescheduledTasks.add(retried);
+        Task taskToBeRetried = failedTask.copy();
+        taskToBeRetried.setTaskId(IDGenerator.generate());
+        taskToBeRetried.setRetriedTaskId(failedTask.getTaskId());
+        taskToBeRetried.setStatus(SCHEDULED);
+        taskToBeRetried.setRetryCount(failedTask.getRetryCount() + 1);
+        rescheduledTasks.add(taskToBeRetried);
+
+        // update the failed task in the DAO
+        failedTask.setRetried(true);
+        executionDAO.updateTask(failedTask);
 
         // Reschedule the cancelled task but if the join is cancelled set that to in progress
-        cancelledTasks.forEach(t -> {
-            if (t.getTaskType().equalsIgnoreCase(WorkflowTask.Type.JOIN.toString())) {
-                t.setStatus(IN_PROGRESS);
-                t.setRetried(false);
-                executionDAO.updateTask(t);
+        cancelledTasks.forEach(task -> {
+            if (task.getTaskType().equalsIgnoreCase(WorkflowTask.Type.JOIN.toString())) {
+                task.setStatus(IN_PROGRESS);
+                executionDAO.updateTask(task);
             } else {
-                //executionDAO.removeTask(t.getTaskId());
-                Task copy = t.copy();
-                copy.setTaskId(IDGenerator.generate());
-                copy.setRetriedTaskId(t.getTaskId());
-                copy.setStatus(SCHEDULED);
-                copy.setRetryCount(t.getRetryCount() + 1);
-                rescheduledTasks.add(copy);
+                Task taskToBeRescheduled = task.copy();
+                taskToBeRescheduled.setTaskId(IDGenerator.generate());
+                taskToBeRescheduled.setRetriedTaskId(task.getTaskId());
+                taskToBeRescheduled.setStatus(SCHEDULED);
+                taskToBeRescheduled.setRetryCount(task.getRetryCount() + 1);
+                rescheduledTasks.add(taskToBeRescheduled);
+                // since the canceled task is being retried, update this
+                task.setRetried(true);
+                executionDAO.updateTask(task);
             }
         });
 
         scheduleTask(workflow, rescheduledTasks);
 
         workflow.setStatus(WorkflowStatus.RUNNING);
         executionDAO.updateWorkflow(workflow);
+        executionDAO.updateTasks(workflow.getTasks());
```

Comment:
Why update the tasks here since the update already happens for each task in the `forEach` ?

Reply:
inside the foreach, only the cancelled tasks are updated. This is applied for all tasks and is used as a data enrichment operation (specific use case is for archived workflows which are being retried).